### DISCUSSION TITLE
Null values

### DISCUSSION THREAD

**Discussion by flying-sheep** [nushell/nushell#5486] [Ideas]
Inspired by https://github.com/nushell/nushell/issues/5469 and https://github.com/nushell/nushell/issues/4975, I’m thinking maybe there should be some thought about null values in general.

We might want more of those, though certainly not all. AFAIK there’s:

- [`mu`](https://en.wikipedia.org/wiki/Mu_%28negative%29): “There is no answer, as the question is invalid”. A common, well defined version of this is:

  - `NaN`: a value that’s supposed to be a number, but due to some failed assumption now holds an invalid value (e.g. through division by zero). This propagates errors through vectorized operations: `[1, NaN] * 2` evaluates to `[2, NaN]` and `max([1, NaN])` evaluates to `NaN`. usually has the weird property that `NaN ≠ NaN` in a comparison, so `[1, NaN] == [1, NaN]` evaluates to `[true, false]`.

- `NA`: missing data, “there is a value but we don’t know”. vector operations over a vector containing NA usually just skip those. so e.g. `max([1, NA])` is 1.
- `null`: absence of a concrete value. can mean many things dependent on context, often it’s both used as a parameter meaning “choose automatically” or “use the default”, and as a result value meaning “no search result”.

I think nushell’s focus on vectorized operations means that we should have a way to emit data that either makes e.g. `some-df | where col > 1` skip values or throw an error.

I think an operation like `ls -l | get size` should be able to return something like `[1kb, 0b, NA]` instead of only `[1kb, 0b, 0b]` like proposed in #5485.

**Comment by webbedspace** [comment]
Doesn't `null` already exist (as a value of type `nothing`)?
```
〉null | describe
nothing
```

**Comment by theAkito** [comment]
I think all the other alternatives aside from `null` you suggested are extremely complicated and would just confuse people coming from different perspectives & languages.

Like...

```
[1, NaN] * 2 evaluates to [2, NaN] and max([1, NaN]) evaluates to NaN
```

Say what? Does not make any sense to me, intuitively. Sure, it has a system, but it's way too complicated and would probably break a lot of code.
Besides that, this example is already more or less covered  by `null`.
